File: library/src/main/java/ke/tang/ruler/ClickableMarker.java
package ke.tang.ruler;

import android.view.View;

public abstract class ClickableMarker implements Marker {
    private OnMarkerClickListener mOnMarkerClickListener;

    private float mX;
    private float mY;

    public ClickableMarker(OnMarkerClickListener onMarkerClickListener) {
        mOnMarkerClickListener = onMarkerClickListener;
    }

    public ClickableMarker() {
    }

    public OnMarkerClickListener getOnMarkerClickListener() {
        return mOnMarkerClickListener;
    }

    public void setOnMarkerClickListener(OnMarkerClickListener onMarkerClickListener) {
        mOnMarkerClickListener = onMarkerClickListener;
    }

    @Override
    public void performClick() {
        if (null != mOnMarkerClickListener) {
            mOnMarkerClickListener.onMarkerClick(this);
        }
    }

    @Override
    public void onAttach(View v) {
        
    }

    public interface OnMarkerClickListener {
        void onMarkerClick(Marker marker);
    }

    @Override
    public float getX() {
        return mX;
    }

    @Override
    public float getY() {
        return mY;
    }

    @Override
    public void setX(float x) {
        mX = x;
    }

    @Override
    public void setY(float y) {
        mY = y;
    }
}

File: library/src/main/java/ke/tang/ruler/ClickableMarker.kt
package ke.tang.ruler

import android.os.Parcel
import java.util.*

abstract class ClickableMarker : Marker {
    override var x: Float = 0f
    override var y: Float = 0f
    internal val cacheKey: String
    var onMarkerClickListener: OnMarkerClickListener? = null
        set(value) {
            forEach {
                if (it.value == cacheKey) {
                    remove(it.key)
                }
            }
            field = value
            if(null != value){
                put(value, cacheKey)
            }
        }

    constructor(onMarkerClickListener: OnMarkerClickListener?) {
        this.onMarkerClickListener = onMarkerClickListener
        cacheKey = UUID.randomUUID().toString()
    }

    constructor(parcel: Parcel) {
        cacheKey = parcel.readString()
        forEach {
            if (it.value == cacheKey) {
                onMarkerClickListener = it.key
            }
        }
    }

    override fun performClick() {
        onMarkerClickListener?.onMarkerClick(this)
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(cacheKey)
    }

    companion object onMarkerClickListenerCaches : WeakHashMap<OnMarkerClickListener, String>()
}